src/coreclr/jit/scev.cpp

```diff
@@ -982,19 +1019,92 @@ Scev* ScalarEvolutionContext::Simplify(Scev* scev)
                 ScevConstant* cns1 = (ScevConstant*)op1;
                 ScevConstant* cns2 = (ScevConstant*)op2;
                 int64_t       newValue;
-                if (binop->TypeIs(TYP_INT))
+                if (genTypeSize(binop->Type) == 4)
                 {
-                    newValue = FoldBinop<int32_t>(binop->Oper, static_cast<int32_t>(cns1->Value),
-                                                  static_cast<int32_t>(cns2->Value));
+                    newValue = FoldBinop<uint32_t>(binop->Oper, static_cast<uint32_t>(cns1->Value),
+                                                   static_cast<uint32_t>(cns2->Value));
                 }
                 else
                 {
-                    assert(binop->TypeIs(TYP_LONG));
-                    newValue = FoldBinop<int64_t>(binop->Oper, cns1->Value, cns2->Value);
+                    assert(genTypeSize(binop->Type) == 8);
+                    newValue = FoldBinop<uint64_t>(binop->Oper, static_cast<uint64_t>(cns1->Value),
+                                                   static_cast<uint64_t>(cns2->Value));
                 }
 
                 return NewConstant(binop->Type, newValue);
             }
+            else if (op2->OperIs(ScevOper::Constant))
+            {
+                ScevConstant* cns2 = (ScevConstant*)op2;
+                if (binop->OperIs(ScevOper::Add, ScevOper::Lsh) && (cns2->Value == 0))
+                {
+                    return op1;
+                }
+
+                if (binop->OperIs(ScevOper::Add))
+                {
+                    // (a + c1) + c2 => a + (c1 + c2)
+                    if (op1->OperIs(ScevOper::Add) && (((ScevBinop*)op1)->Op2->OperIs(ScevOper::Constant)))
+                    {
+                        ScevBinop* newOp2 = NewBinop(ScevOper::Add, ((ScevBinop*)op1)->Op2, cns2);
+                        ScevBinop* newAdd = NewBinop(ScevOper::Add, ((ScevBinop*)op1)->Op1, newOp2);
+                        return Simplify(newAdd);
+                    }
+                }
+
+                if (binop->OperIs(ScevOper::Mul))
+                {
+                    // a * 0 => 0
+                    if (cns2->Value == 0)
+                    {
+                        return cns2;
+                    }
+
+                    // a * 1 => 1
+                    if (cns2->Value == 1)
+                    {
+                        return op1;
+                    }
+
+                    // (a + b) * c2 => a * c2 + b * c2
+                    if (op1->OperIs(ScevOper::Add))
+                    {
+                        ScevBinop* newOp1 = NewBinop(ScevOper::Mul, ((ScevBinop*)op1)->Op1, cns2);
+                        ScevBinop* newOp2 = NewBinop(ScevOper::Mul, ((ScevBinop*)op1)->Op2, cns2);
+                        ScevBinop* newAdd = NewBinop(ScevOper::Add, newOp1, newOp2);
+                        return Simplify(newAdd);
+                    }
```

Comment:
This rule seems a bit questionable, but I needed it to simplify something at some point. Might look to see how to get rid of this in a follow up.

Reply:
In the end I just removed it in this PR, doesn't seem to have much impact on the diffs.